### DISCUSSION TITLE
tRPC Server Actions

### DISCUSSION THREAD

**Discussion by juliusmarminge** [trpc/trpc#5737] [Blogs]
Ref: https://trpc.io/blog/trpc-actions

Let us know what you think of these new primitives. Have you already tried them out? What is missing for you to adopt this?

**Comment by cipriancaba** [comment]
I think this looks great, giving it a spin now. 

I kind of like the hooks from the next-safe-action lib https://next-safe-action.dev/docs/usage-from-client/hooks/useaction

Not everything is a server component, so those are nice for more advanced client components

**Reply by juliusmarminge** [reply]
What does the hook provide that `useActionState` doesnt? 

It's likely there'll have to be wrappers around the built-in hooks (especially to fix the type-mismatch) but it would be nice to know what issues people run into.

At work we mostly call actions inline and handle pending state manually (was before useFormState/useActionState) but in trellix-trpc i used useActionState a bit (e.g https://github.com/juliusmarminge/trellix-trpc/blob/e68424370a07b370842883177898854ccc61508a/src/app/boards/%5Bid%5D/column.tsx#L54)

**Reply by cipriancaba** [reply]
Isn't useActionState meant for form actions though? Meaning we would have to wrap the buttons in forms?

As in for the inline action calls, I feel I always want to show some kind of feedback or error after an action, and just having this nicely wrapped in a hook helps me deal with it consistently in a nice declarative way

```
const sendMagicLink = useAction(sendMagicLinkAction, {
    onSuccess: () => {
      setCurrentStep('check-email')
    },
    onError: (e) => {
      toast.error(e.serverError)
    },
  })
```

**Reply by melkir** [reply]
> What does the hook provide that useActionState doesnt?

According to this [blog post](https://medium.com/@michal-worwag/forms-with-next-js-and-react-19-with-usestateaction-35674f734a8d), if an error occurs while submitting the form, the field values are now preserved. Previously, the form would reset, which could be quite frustrating.

**Comment by ally-ahmed** [comment]
Great work! I have been looking into how to use server actions together with routes. I am still experimenting but would the snippet below be the way to go? What do you think? In the snippet below I am trying to use the T3-stack setup but with also server actions. I have done basic tests and it is working well. My use case is to use the server actions for auth. I'm using [trellix-trpc](https://github.com/juliusmarminge/trellix-trpc), [time-report](https://github.com/juliusmarminge/time-report) and [trpc-rq-prefetching](https://github.com/juliusmarminge/trpc-rq-prefetching) as examples on how to use the new experimental features they have been a good resource. Thank you.

```ts
// trpc.ts
const serverActionProcedure = t.procedure.experimental_caller(
  experimental_nextAppDirCaller({
    async createContext() {
      const heads = new Headers(headers());
      heads.set("x-trpc-source", "server-action");
      return createTRPCContext({
        headers: heads,
        session: await auth(),
      });
    },
  }),
);
const publicAction = serverActionProcedure;
const protectedAction = serverActionProcedure.use(({ ctx, next }) => {
  if (!ctx.auth?.user) {
    throw new TRPCError({ code: "UNAUTHORIZED" });
  }
  return next({
    ctx: {
      session: { ...ctx.session, user: ctx.session.user },
    },
  });
});
```

**Comment by beaupd** [comment]
Awesome work guys, one problem when I try to use a component that uses cookies from "next/headers" on mutation it breaks, but on query it works. Super weird. But very blocking. It doesn't seem like im doing anything wrong however if this isn't something anyone else is experience or able to reproduce. I can create a isolated reproduction.

this is the full error:

**"You're importing a component that needs next/headers. That only works in a Server Component which is not supported in the pages/ directory."**

It seems its somehow routing mutation through pages instead of server action.

**Reply by juliusmarminge** [reply]
please provide a reproduction 🙏

**Reply by beaupd** [reply]
I'm not able to make a small isolated replication, its getting weirder. I now fixed it in my actual project, by adding "use server" directive in the imported component file that uses the cookies from "next/headers". The imported function is sync, but i can only make it work by awaiting this synchronous imported function with the "use server" directive. 

However weird this is, I think its a bug for the team at NextJS so I don't think this should be continued here. I'm going to leave it at this for now. I will try and make the bug work in an isolated environment later. But I don't have time for it right now

**Comment by jonsoku-dev** [comment]
I hope you are all doing well. I have a question regarding `server action with trpc`. 
I would appreciate any guidance or insights you can provide.


```tsx
import { initTRPC, TRPCError } from '@trpc/server';
import { experimental_nextAppDirCaller } from '@trpc/server/adapters/next-app-dir';
import { currentUser } from './auth';
 
interface Meta {
  span: string;
}
 
export const t = initTRPC.meta<Meta>().create();
 
export const serverActionProcedure = t.procedure
  .experimental_caller(
    experimental_nextAppDirCaller({
      pathExtractor: ({ meta }) => (meta as Meta).span,
    }),
  )
  .use(async (opts) => {
    // Inject user into context
    const user = await currentUser();
           
const user: User | null
    return opts.next({ ctx: { user } });
  });
 
export const protectedAction = serverActionProcedure.use((opts) => {
  if (!opts.ctx.user) {
    throw new TRPCError({
      code: 'UNAUTHORIZED',
    });
  }
 
  return opts.next({
    ctx: {
      ...opts.ctx,
      user: opts.ctx.user, // <-- ensures type is non-nullable
       
(property) user: User
    },
  });
});
```

```tsx
// app/_actions.tsx
'use server';
 
import { z } from 'zod';
import { protectedAction } from '../server/trpc';
 
export const createPost = protectedAction
  .input(
    z.object({
      title: z.string(),
    }),
  )
  .mutation(async ({ ctx, input }) => {
    // Do something with the input
  });
 
// Since we're using the `experimental_caller`,
// our procedure is now just an ordinary function:
createPost;


// if add this query
export const userList = protectedAction
  .input(
    z.undefined(),
  )
  .query(async ({ ctx, input }) => {
    // Do something with the input
  });

userList;
```

The code above shows the server action part of the documentation. How do I integrate it with the client part using the existing TanStack React Query? Do I need to write both the query in the server action and the query in React Query separately? I would like to know how to share the router, similar to the createCaller type.

ref: https://trpc.io/docs/quickstart#2-add-a-query-procedure

```tsx
import { db } from './db';
import { publicProcedure, router } from './trpc';
 
const appRouter = router({
  userList: publicProcedure
    .query(async () => {
      // Retrieve users from a datasource, this is an imaginary database
      const users = await db.user.findMany();
             
const users: User[]
      return users;
    }),
});

export const createCaller = createCallerFactory(appRouter);

export const trpcServer = createCaller(createContext, {
   ...
});


// app/some-page/page.tsx
const Page = async ({ searchParams }: PageProps) => {
  const data = await trpcServer.userList();

  return (
    <>
      ...
    </>
  );
};

export default Page
```

Sorry for the disorganized question. If my question is unclear, please focus on the following:

I would like to know if there is a way to write a router once, similar to trpcServer, and use it simultaneously on both the server and client. It seems there are no examples of this in the current trpc server action documentation. Is this possible?
Or should server actions be considered separately?

Thank you in advance for your help!

**Reply by beaupd** [reply]
Correct me if I'm wrong, but since you use 2 completely different procedures and the server action procedure(experimental_nextAppDirCaller) is not backwards compatible as of right now, you need at least 2 different procedures (publicProcedure and protectedAction). I don't think it's possible at this moment to do it clean and neat since its all very experimental, however you can reuse the function and the (Zod) schema. Which I did in my project, it's not perfect but its not too bad since the actual functionality and its types is in one place. 

Hope that helped a little :)

**Reply by juliusmarminge** [reply]
actions are not bound to a router. you could import it on the client and call it as your queryFn though if you want to use it to fetch data

**Comment by dBianchii** [comment]
Why can I use .query() inside a trpc server action? Since server actions only support POST requests

**Reply by juliusmarminge** [reply]
You can't if you call your action from `useActionState.dispatch`.

Only mutations will handle formdata for example, and pull the input from the 2nd argument.

See here for the technical difference https://github.com/trpc/trpc/blob/e805970e83aa71e642601feb864848003ef0e3b9/packages/server/src/adapters/next-app-dir/nextAppDirCaller.ts#L78

You can use `.query` as a data access layer

**Comment by codesfromshad** [comment]
@juliusmarminge getting `Cannot read properties of undefined (reading 'span')` for some reason.

**Reply by juliusmarminge** [reply]
Do you have `meta` set?

**Reply by codesfromshad** [reply]
Yes, on the T3 Stack.

```TypeScript
interface Meta extends TRPCPanelMeta {
  span: string;
}

/**
 * 2. INITIALIZATION
 *
 * This is where the tRPC API is initialized, connecting the context and transformer. We also parse
 * ZodErrors so that you get typesafety on the frontend if your procedure fails due to validation
 * errors on the backend.
 */
const t = initTRPC.context<typeof createTRPCContext>().meta<Meta>().create({
  transformer: SuperJSON,
  errorFormatter({ shape, error }) {
    return {
      ...shape,
      data: {
        ...shape.data,
        zodError:
          error.cause instanceof ZodError ? error.cause.flatten() : null,
      },
    };
  },
});
```

**Reply by juliusmarminge** [reply]
That's just the type, I meant runtime meta on each procedure?

**Reply by codesfromshad** [reply]
What do I pass in meta for each procedure? According to [this](https://github.com/trpc/trpc/discussions/5737#discussioncomment-9935162:~:text=export%20const%20t%20%3D,currentUser()%3B), what did I do wrong?

**Reply by juliusmarminge** [reply]
See https://trpc.io/blog/trpc-actions#observability

```ts
export const createPost = protectedAction
  .meta({ span: 'createPost' }) // <---  
  .input(
    z.object({
      title: z.string(),
    }),
  )
  .mutation(async ({ ctx, input }) => {
    // Do something with the input
  });
```

If you don't care about that then you can also skip the `pathExtractor` argument 

https://trpc.io/blog/trpc-actions#defining-server-actions-with-trpc

```ts
export const t = initTRPC.meta<Meta>().create();
 
const serverActionProcedure = t.procedure.experimental_caller(
  experimental_nextAppDirCaller({
    // pathExtractor: ({ meta }) => (meta as Meta).span,
  }),
);
```

**Reply by codesfromshad** [reply]
Still wasn't able to get it working, can you share your repo where you got it working?

**Reply by juliusmarminge** [reply]
It's linked at the end of the article

**Reply by codesfromshad** [reply]
Found it, thanks!

**Comment by meslzy** [comment]
Well done. its amazing. but, how can i use `errorFormatter` seems not working.

```ts
export const t = initTRPC.meta<Meta>().create({
  errorFormatter: () => {
    return {
      code: "INTERNAL_SERVER_ERROR",
      message: "Internal server error",
    };
  },
});
```

the error showing in console due validation error 

```txt
digest: "3673864806"
message: "3 errors occurred"
stack: "TRPCError: 3 errors occurred\n    at inputValidatorMiddleware , ..etc"
```

**Comment by gonta1026** [comment]
When I specify createPost as action in the link below, I assign `type '(input: { title: string; }) => Promise<void>' to type 'string | ((formData: FormData) => void) | undefined'. The `create_promise_type' error occurs.
 I want to modify createPost to match the type `(formData: FormData) => void`, how can I do this?
 
https://trpc.io/blog/trpc-actions#defining-server-actions-with-trpc


```tsx
'use client'
import { createPost } from '../_actions'
export default function PostForm() {
  // }
  return (
    <form action={createPost}>
      {/* error: Type '(input: { title: string; }) => Promise<void>' cannot be assigned to type 'string | ((formData: FormData) => void) | undefined'.
       */}
      <input type="text" name="title" />
      <button type="submit">Create Post</button>
    </form>
  )
}

```

node: v20.5.0
package:
```
  "dependencies": {
    "@baselime/node-opentelemetry": "^0.5.8",
    "@prisma/client": "^5.16.2",
    "@tanstack/react-query": "^5.51.11",
    "@trpc/client": "^11.0.0-rc.466",
    "@trpc/next": "^11.0.0-rc.466",
    "@trpc/react-query": "^11.0.0-rc.466",
    "@trpc/server": "^11.0.0-rc.467",
    "next": "14.2.5",
    "react": "^18",
    "react-dom": "^18",
    "superjson": "^2.2.1",
    "zod": "^3.23.8"
  },
```

**Reply by beaupd** [reply]
I dont think You can change the formdata type, you just have to parse it client side and server side with a schema like zod schema, also check synchronous or asynchronous functions.

**Reply by JonParton** [reply]
I came across the same thing experimenting today ... I ended up having to just mark the server action as `any` to ignore the issue. 

```ts
import { addNotification } from "./_actions";

...

<form
   // Unfortunately Form Data is an untyped mess ... So we have to use any if we want to be able to pass the data without javascript 🤦
   // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
   action={addNotification as any}
>
   {...}
</form>
```

At run time it works just fine and I noticed in Julius's Trellix repo he also had to do quite a few any's and similar. 

I know you could parse the form data on the client side in JavaScript, but that breaks the way you want to make it progressive being able to pass it to the `form.action` prop so it works without JavaScript 🤔

Wonder if the implementation of the `t.procedure.experimental_caller({...})` could be changed to accept either the actual input schema type or `FormData` so this kind of use gets ignored by TS? 🤔

**Comment by jisunglim** [comment]
I read your blog post (https://trpc.io/blog/trpc-actions) and found it very informative. However, since the example isn’t framework-agnostic, I looked for other examples in the tRPC repository.

In another example, I noticed that using `experimental_createServerActionHandler` allows us to convert each existing framework-agnostic procedure into a Server Action, one by one. However, in this case, while individual procedures remain framework-agnostic, there doesn’t seem to be a way to create Server Actions directly from a router (which consists of multiple procedures) while keeping the router itself framework-agnostic.

In this situation, similar to `createCallerFactory`, how about introducing something like `createNextServerActionFactory` to enable us to create Server Actions directly from a router? For example:

### `app/server/trpc.ts`
**Basic initialization and setup of tRPC**
```ts
const t = initTRPC.context<Context>().create({
  //...
});

export const router = t.router;
export const publicProcedure = t.procedure;

export const createNextServerActionFactory = t.createNextServerActionFactory;

// ...
```

### `app/server/routers/_app.ts`

**Implementation of procedures and routers (DRY)**
_Up to this point, everything is framework-agnostic_
```ts
import { publicProcedure, router } from '~/server/trpc';

export const createPost = publicProcedure
  .input(
    z.object({
      title: z.string(),
      content: z.string(),
    }),
  )
  .mutation(async (opts) => {
    latestPost = {
      id: latestPost.id + 1,
      createdAt: Temporal.Now.plainDateTimeISO('Asia/Singapore'),
      ...opts.input,
    };

    return latestPost;
  });
  
export const appRouter = router({
  createPost,
  // ...
});

export type AppRouter = typeof appRouter;
```

### `app/trpc/server-action.ts`
**Framework-specific code starts here (Next.js Server Actions)**
_Instead of creating Server Actions for each individual procedure, we convert the entire router into Server Actions_
```ts
import { appRouter } from '~/server/routers/_app';
import { createNextServerActionFactory } from '~/server/trpc';
//...

const createContext = cache(async (): Promise<Context> => {
  //...
});

const createNextServerAction = createNextServerActionFactory(appRouter);
export const nextServerAction = createNextServerAction(createContext);
```

### `app/post-form.tsx`
**Accessing Server Actions in the same way as procedures in tRPC**
```tsx
'use client';

import { nextServerAction } from '~/_actions';

export function PostForm() {
  return (
    <form
      // Use `action` to make form progressively enhanced
      action={nextServerAction.createPost}
      // `Using `onSubmit` allows building rich interactive
      // forms once JavaScript has loaded
      onSubmit={async (e) => {
        e.preventDefault();
        const title = new FormData(e.target).get('title');
        // Maybe show loading toast, etc etc. Endless possibilities
        await nextServerAction.createPost({ title });
      }}
    >
      <input type="text" name="title" />
      <button type="submit">Create Post</button>
    </form>
  );
}
```

By doing this, we maintain a framework-agnostic approach in our procedures and routers and only introduce framework-specific code when converting the router into Server Actions.

What do you think about this approach? Is there any plan to introduce something like createNextServerActionFactory to the tRPC library to facilitate this pattern?

**Reply by jblevins1991** [reply]
This should never be bound directly to trpc's api though. It would have to be moved to the nextjs adapter to be namespaced and so that trpc is kept framework agnostic.
